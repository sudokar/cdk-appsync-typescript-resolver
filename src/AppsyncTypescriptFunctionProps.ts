// ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
import type { aws_appsync } from 'aws-cdk-lib';

/**
 * AppsyncTypescriptFunctionProps
 */
export interface AppsyncTypescriptFunctionProps {
  /**
   * the data source linked to this AppSync Function.
   * @stability stable
   */
  readonly dataSource: aws_appsync.BaseDataSource;
  /**
   * the GraphQL Api linked to this AppSync Function.
   * @stability stable
   */
  readonly api: aws_appsync.IGraphqlApi;
  /**
   * the response mapping template for the AppSync Function.
   * @default - no response mapping template
   * @stability stable
   */
  readonly responseMappingTemplate?: aws_appsync.MappingTemplate;
  /**
   * the request mapping template for the AppSync Function.
   * @default - no request mapping template
   * @stability stable
   */
  readonly requestMappingTemplate?: aws_appsync.MappingTemplate;
  /**
   * the description for this AppSync Function.
   * @default - no description
   * @stability stable
   */
  readonly description?: string;
  /**
   * the name of the AppSync Function.
   * @stability stable
   */
  readonly name: string;
  /**
   * A map of replacement strings in the bundled code. e.g { ENV: "PROD" }
   */
  readonly replaceStrings?: Record<string, string>;
  /**
   * Flag to enable or disable source maps in bundled code. defaults to false
   */
  readonly sourceMap?: boolean;
  /**
   * Path of typescript file that will be transpiled and bundled
   */
  readonly path: string;
}
