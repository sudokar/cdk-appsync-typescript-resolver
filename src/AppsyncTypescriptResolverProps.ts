// ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
import { aws_appsync } from 'aws-cdk-lib';

/**
 * AppsyncTypescriptResolverProps
 */
export interface AppsyncTypescriptResolverProps {
  /**
   * The API this resolver is attached to.
   * @stability stable
   */
  readonly api: aws_appsync.IGraphqlApi;
  /**
   * The data source this resolver is using.
   * @default - No datasource
   * @stability stable
   */
  readonly dataSource?: aws_appsync.BaseDataSource;
  /**
   * configuration of the pipeline resolver.
   * @default - no pipeline resolver configuration
An empty array | undefined sets resolver to be of kind, unit
   * @stability stable
   */
  readonly pipelineConfig?: Array<aws_appsync.IAppsyncFunction>;
  /**
   * The maximum number of elements per batch, when using batch invoke.
   * @default - No max batch size
   * @stability stable
   */
  readonly maxBatchSize?: number;
  /**
   * The caching configuration for this resolver.
   * @default - No caching configuration
   * @stability stable
   */
  readonly cachingConfig?: aws_appsync.CachingConfig;
  /**
   * name of the GraphQL type this resolver is attached to.
   * @stability stable
   */
  readonly typeName: string;
  /**
   * name of the GraphQL field in the given type this resolver is attached to.
   * @stability stable
   */
  readonly fieldName: string;
  /**
   * A map of replacement strings in the bundled code. e.g { ENV: "PROD" }
   */
  readonly replaceStrings?: Record<string, string>;
  /**
   * Flag to enable or disable source maps in bundled code. defaults to false
   */
  readonly sourceMap?: boolean;
  /**
   * Path of typescript file that will be transpiled and bundled
   */
  readonly path?: string;
}
