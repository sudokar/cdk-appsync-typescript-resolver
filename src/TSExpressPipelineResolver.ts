import * as appsync from 'aws-cdk-lib/aws-appsync';
import type { IConstruct } from 'constructs';
import type { TSExpressPipelineResolverProps } from './TSExpressPipelineResolverProps';

const defaultPipelineCode: string = `
// The before step
export function request() {
    return {}
}
// The after step
export function response(ctx) {
    return ctx.prev.result
}`;

/**
 * Appsync's JS pipeline resolver with default bolierplate code using AppsyncTypescriptFunction construct
 * @deprecated - Appsync now natively supports Unit resolvers. Ref: https://aws.amazon.com/about-aws/whats-new/2023/08/aws-appsync-javascript-all-resolvers-graphql-apis
 */
export class TSExpressPipelineResolver extends appsync.Resolver {
  constructor(
    scope: IConstruct,
    id: string,
    props: TSExpressPipelineResolverProps,
  ) {
    const { tsFunction: appsyncFunction, ...resolverProps } = props;
    super(scope, id, {
      ...resolverProps,
      pipelineConfig: [appsyncFunction],
      code: appsync.Code.fromInline(defaultPipelineCode),
      runtime: appsync.FunctionRuntime.JS_1_0_0,
    });
  }
}
