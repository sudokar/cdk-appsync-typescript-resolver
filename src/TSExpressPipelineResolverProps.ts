// ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
import type { aws_appsync } from 'aws-cdk-lib';
import type { AppsyncTypescriptFunction } from './';

/**
 * TSExpressPipelineResolverProps
 */
export interface TSExpressPipelineResolverProps {
  /**
   * The API this resolver is attached to.
   * @stability stable
   */
  readonly api: aws_appsync.IGraphqlApi;
  /**
   * The data source this resolver is using.
   * @default - No datasource
   * @stability stable
   */
  readonly dataSource?: aws_appsync.BaseDataSource;
  /**
   * The maximum number of elements per batch, when using batch invoke.
   * @default - No max batch size
   * @stability stable
   */
  readonly maxBatchSize?: number;
  /**
   * The caching configuration for this resolver.
   * @default - No caching configuration
   * @stability stable
   */
  readonly cachingConfig?: aws_appsync.CachingConfig;
  /**
   * name of the GraphQL type this resolver is attached to.
   * @stability stable
   */
  readonly typeName: string;
  /**
   * name of the GraphQL field in the given type this resolver is attached to.
   * @stability stable
   */
  readonly fieldName: string;
  /**
   * Instance of AppsyncTypescriptFunction construct
   */
  readonly tsFunction: AppsyncTypescriptFunction;
}
