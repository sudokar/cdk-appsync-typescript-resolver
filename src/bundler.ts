import * as esbuild from 'esbuild';

type BundleProps = Readonly<{
  entryPoint: string;
  sourceMap: boolean;
  replaceStrings?: { [k: string]: string };
}>;

export const bundle = (props: BundleProps) => {
  const { entryPoint, sourceMap, replaceStrings } = props;

  const buildResult = esbuild.buildSync({
    entryPoints: [entryPoint],
    external: ['@aws-appsync/utils'],
    bundle: true,
    write: false,
    platform: 'node',
    target: 'esnext',
    format: 'esm',
    sourcemap: sourceMap ? 'inline' : false,
    sourcesContent: false,
  });

  let bundledCode = buildResult.outputFiles[0].text;

  if (replaceStrings) {
    Object.entries(replaceStrings).forEach(([key, value]) => {
      const regExp = new RegExp(key, 'g');
      bundledCode = bundledCode.replace(regExp, value);
    });
  }

  return bundledCode;
};

export const defaultResolverCode: string = `
// The before step
export function request() {
    return {}
}
// The after step
export function response(ctx) {
    return ctx.prev.result
}`;
