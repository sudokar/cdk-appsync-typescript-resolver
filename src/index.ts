export * from './AppsyncTypescriptFunction';
export * from './AppsyncTypescriptFunctionProps';
export * from './TSExpressPipelineResolver';
export * from './TSExpressPipelineResolverProps';
export * from './TypescriptUnitResolver';
export * from './TypescriptUnitResolverProps';
