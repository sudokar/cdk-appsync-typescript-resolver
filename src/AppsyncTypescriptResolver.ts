import * as appsync from 'aws-cdk-lib/aws-appsync';
import type { IConstruct } from 'constructs';
import type { AppsyncTypescriptResolverProps } from './AppsyncTypescriptResolverProps';
import { bundle, defaultResolverCode } from './bundler';

/**
 * Transpile and bundle Typescript to AWS Appsync JS resolver
 */
export class AppsyncTypescriptResolver extends appsync.Resolver {
  constructor(
    scope: IConstruct,
    id: string,
    props: AppsyncTypescriptResolverProps,
  ) {
    const { path, sourceMap, replaceStrings, ...resolverProps } = props;
    super(scope, id, {
      ...resolverProps,
      code: appsync.Code.fromInline(
        path ? bundle({
          entryPoint: path,
          sourceMap: sourceMap ?? false,
          replaceStrings,
        }) : defaultResolverCode,
      ),
      runtime: appsync.FunctionRuntime.JS_1_0_0,
    });
  }
}
