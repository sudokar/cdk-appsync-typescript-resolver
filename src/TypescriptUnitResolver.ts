import * as appsync from 'aws-cdk-lib/aws-appsync';
import type { IConstruct } from 'constructs';
import { bundle } from './bundler';
import type { TypescriptUnitResolverProps } from './TypescriptUnitResolverProps';

export class TypescriptUnitResolver extends appsync.Resolver {
  constructor(scope: IConstruct, id: string, props: TypescriptUnitResolverProps) {
    const {
      path,
      sourceMap,
      replaceStrings,
      ...resolverProps
    } = props;

    super(scope, id, {
      code: appsync.Code.fromInline(
        bundle({
          entryPoint: path,
          sourceMap: sourceMap ?? false,
          replaceStrings,
        }),
      ),
      runtime: appsync.FunctionRuntime.JS_1_0_0,
      ...resolverProps,
    });
  }
}
