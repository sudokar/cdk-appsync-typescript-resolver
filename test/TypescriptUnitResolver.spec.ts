import * as path from 'path';
import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as appsync from 'aws-cdk-lib/aws-appsync';
import {
  getTestApi,
  getTestStack,
} from './testUtils';
import { TypescriptUnitResolver } from '../src';

let testStack: cdk.Stack;
let testApi: appsync.GraphqlApi;

describe('TypescriptUnitResolver', () => {
  beforeEach(() => {
    testStack = getTestStack();
    testApi = getTestApi(testStack);
  });

  test('should create JS unit resolver for typescript file path', () => {
    new TypescriptUnitResolver(testStack, 'TestResolver', {
      api: testApi,
      typeName: 'Query',
      fieldName: 'hello',
      path: path.join(__dirname, 'resources', 'getTask.ts'),
    });

    const template = Template.fromStack(testStack);
    template.hasResourceProperties('AWS::AppSync::Resolver', {
      TypeName: 'Query',
      FieldName: 'hello',
      Runtime: {
        Name: 'APPSYNC_JS',
        RuntimeVersion: '1.0.0',
      },
    });
    expect(template).toMatchSnapshot();
  });
});
