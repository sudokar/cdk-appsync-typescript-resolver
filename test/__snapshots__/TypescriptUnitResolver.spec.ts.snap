// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypescriptUnitResolver should create JS unit resolver for typescript file path 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ApiDefaultApiKeyF991C37B": {
      "DependsOn": [
        "ApiSchema510EECD7",
      ],
      "Properties": {
        "ApiId": {
          "Fn::GetAtt": [
            "ApiF70053CD",
            "ApiId",
          ],
        },
      },
      "Type": "AWS::AppSync::ApiKey",
    },
    "ApiF70053CD": {
      "Properties": {
        "AuthenticationType": "API_KEY",
        "Name": "test-api",
      },
      "Type": "AWS::AppSync::GraphQLApi",
    },
    "ApiSchema510EECD7": {
      "Properties": {
        "ApiId": {
          "Fn::GetAtt": [
            "ApiF70053CD",
            "ApiId",
          ],
        },
        "Definition": "type Task {
    id: ID!
    title: String!
    completed: Boolean!
    user: User!
}

type User {
    username: String!
    name: String
    tasks: [Task]
}

type Query {
    getTask(id: ID!): Task
}",
      },
      "Type": "AWS::AppSync::GraphQLSchema",
    },
    "TestResolver0032C08D": {
      "DependsOn": [
        "ApiSchema510EECD7",
      ],
      "Properties": {
        "ApiId": {
          "Fn::GetAtt": [
            "ApiF70053CD",
            "ApiId",
          ],
        },
        "Code": "// test/resources/getTask.ts
import { util as util2 } from "@aws-appsync/utils";

// test/resources/utils.ts
import { util } from "@aws-appsync/utils";
var generateAutoId = () => util.autoId();

// test/resources/getTask.ts
function request(ctx) {
  console.log("ctx.args.id", ctx.args.id);
  return {
    operation: "GetItem",
    key: {
      id: util2.dynamodb.toDynamoDB(ctx.args.id)
    }
  };
}
function response(ctx) {
  return { ...ctx.result, id: generateAutoId() };
}
export {
  request,
  response
};
",
        "FieldName": "hello",
        "Kind": "UNIT",
        "Runtime": {
          "Name": "APPSYNC_JS",
          "RuntimeVersion": "1.0.0",
        },
        "TypeName": "Query",
      },
      "Type": "AWS::AppSync::Resolver",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
